torch/cuda/comm.py

```diff
@@ -51,34 +52,38 @@ def reduce_add(inputs, destination=None):
 
     Returns:
         A tensor containing an elementwise sum of all inputs, placed on the
-        ``destination`` device.
+        :attr:`destination` device.
     """
-    # TODO: try to find an input on another gpu, copy it,
-    # and accumulate into the copy
-    if destination is None:
-        destination = torch.cuda.current_device()
+    destination = torch.cuda._utils._get_device_index(destination, optional=True)
     input_size = inputs[0].size()
-    nccl_root = None
+    root_index = None  # index of input tensor that already is on the correct device
     for i, inp in enumerate(inputs):
         assert inp.is_cuda, "reduce_add expects all inputs to be on GPUs"
         if inp.get_device() == destination:
-            nccl_root = i
+            root_index = i
         if inp.size() != input_size:
             got = 'x'.join(str(x) for x in inp.size())
             expected = 'x'.join(str(x) for x in input_size)
             raise ValueError("input {} has invalid size: got {}, but expected "
                              "{}".format(i, got, expected))
-    if nccl_root is None:
+    if root_index is None:
         raise RuntimeError("reduce_add expects destination to be on the same GPU with one of the tensors")
-    result = inp.new(device=destination).resize_as_(inp).zero_()
-
-    if nccl.is_available(inputs) and inputs[0].get_device() == destination:
-        outputs = [result] + [t.new(t.size()) for t in inputs[1:]]
-        nccl.reduce(inputs, outputs, root=nccl_root)
-        return result
-    for inp in inputs:
-        input_correct_gpu = inp.cuda(result.get_device())
-        result.add_(input_correct_gpu)
+
+    if len(inputs) == 1:
+        return inputs[0]
+
+    if nccl.is_available(inputs):
+        result = torch.empty_like(inputs[root_index])
+        nccl.reduce(inputs, output=result, root=root_index)
+    else:
+        # clone inputs[root_index] and accuimulate into the copy
+        nonroot = [t for i, t in enumerate(inputs) if i != root_index]
+        result = inputs[root_index]
+        for i, other in enumerate(nonroot):
+            if i == 0:
+                result = result.add(other.cuda(destination, non_blocking=True))  # make a new tensor
```

Comment:
Can we just clone `inputs[root_index]` to `result` out of the loop and save a branch here?